src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -304,22 +381,31 @@ impl NailgunProcess {
       &startup_options.argv[1..],
       &workdir_path
     );
-    let handle = std::process::Command::new(&cmd)
-      .args(&startup_options.argv[1..])
-      .stdout(Stdio::piped())
-      .stderr(Stdio::piped())
-      .current_dir(&workdir_path)
-      .spawn();
-    handle
-      .map_err(|e| {
-        format!(
-          "Failed to create child handle with cmd: {} options {:#?}: {}",
-          &cmd, &startup_options, e
-        )
+    Self::create_output_file(workdir_path.clone(), "stdout.log".to_string())
+      .join(Self::create_output_file(
+        workdir_path.clone(),
+        "stderr.log".to_string(),
+      ))
```

Comment:
I couldn't figure out a way to turn a Stdio::piped object into something writable to a file easily so I went the other direction. It does mean that we have a poll the filesystem a couple times, but I haven't ever seen it take more than 2 loops. My guess is that the previous way of .read() ing the stdout pipe was blocking, but now we are using AsyncRead on the file handle we don't get the impression that things just happen right away. There is a little overhead for flushing to disk, but its negligible compared to the lifetime of the server.

Reply:
I agree, given that this also has the benefit of always ensuring that we get the full log, because it's not this process that's writing the logs anywhere. So all in all, it looks good.

Small nit though, could you move the string filename to a constant?

Reply:
Can you describe what this would _ideally_ look like not going via the disk? Maybe I can pull out a pile of types to string together to do that (maybe not!) :)

Reply:
In my mind, the ideal way to model this is as a `Stream`, where the elements are futures of lines. In that case, we can say something like `myStdoutStream.next().and_then(|line_containing_port| {// do whatever})`, and we don't have to poll, we let the futures polling do it for us. I'm not even sure that's possible or that it makes sense in the `Futures` model, but it would be neat.

We'd then have to find a way to pipe the rest of the log to a file, because we still want to keep it. Since we'd still need to pipe everything to a file regardless, I'm not sure how useful that `Stream` would be.

However! Can't we read the stdout file as a stream? Open it asynchronously, as that `Stream<Future<Line>>`?

Reply:
So ideally we'd like a type which goes:
I have a Stdio::piped() output
I will tee this to two places. The first is just to write straight to a file, the second is a thing that will peek at the output, and drive a `Future` which will resolve to some value based on reading the output line by line?

Reply:
Yep, that would be ideal in my head, but @hrfuller might have a different idea :)

Reply:
I agree that is the ideal way to to do this, but I couldn't fine a tee-ing primative, and to do the tee-ing manually would mean holding on to the stdio piped handle and dealing with writing all the lines to a file. Its much simpler to have the process framework handle that.

Reply:
I will change the timings for a tighter loop, as daniel suggested.